diff --git a/org/apache/jasper/JspCompilationContext.java b/org/apache/jasper/JspCompilationContext.java
index 9f602be3207a..1c6173e8ff62 100644
--- a/org/apache/jasper/JspCompilationContext.java
+++ b/org/apache/jasper/JspCompilationContext.java
@@ -688,12 +688,12 @@ public class JspCompilationContext {
         // Append servlet or tag handler path to scratch dir
         try {
             File base = options.getScratchDir();
-            baseUrl = base.toURI().toURL();
             outputDir = base.getAbsolutePath() + File.separator + path +
                     File.separator;
             if (!makeOutputDir()) {
                 throw new IllegalStateException(Localizer.getMessage("jsp.error.outputfolder"));
             }
+            baseUrl = base.toURI().toURL();
         } catch (MalformedURLException e) {
             throw new IllegalStateException(Localizer.getMessage("jsp.error.outputfolder"), e);
         }
