diff --git a/org/apache/jasper/JspCompilationContext.java b/org/apache/jasper/JspCompilationContext.java
index 1c6173e8ff62..8c23762f8dcd 100644
--- a/org/apache/jasper/JspCompilationContext.java
+++ b/org/apache/jasper/JspCompilationContext.java
@@ -179,7 +179,8 @@ public class JspCompilationContext {
             jspLoader = new JasperLoader
                     (new URL[] {baseUrl},
                             getClassLoader(),
-                            rctxt.getPermissionCollection());
+                            rctxt.getPermissionCollection(),
+                            rctxt.getCodeSource());
         }
         return jspLoader;
     }
diff --git a/org/apache/jasper/servlet/JasperLoader.java b/org/apache/jasper/servlet/JasperLoader.java
index 4f379392f2a2..cba48a09b28c 100644
--- a/org/apache/jasper/servlet/JasperLoader.java
+++ b/org/apache/jasper/servlet/JasperLoader.java
@@ -17,14 +17,19 @@
 
 package org.apache.jasper.servlet;
 
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.URL;
 import java.net.URLClassLoader;
+import java.security.AccessController;
 import java.security.CodeSource;
 import java.security.PermissionCollection;
+import java.security.PrivilegedAction;
+import java.security.ProtectionDomain;
 
 import org.apache.jasper.Constants;
+import org.apache.jasper.security.SecurityUtil;
 
 /**
  * Class loader for loading servlet class files (corresponding to JSP files)
@@ -37,12 +42,15 @@ public class JasperLoader extends URLClassLoader {
 
     private final PermissionCollection permissionCollection;
     private final SecurityManager securityManager;
+    private final CodeSource _codeSource;
 
     public JasperLoader(URL[] urls, ClassLoader parent,
-                        PermissionCollection permissionCollection) {
+                        PermissionCollection permissionCollection,
+                        CodeSource codeSource) {
         super(urls, parent);
         this.permissionCollection = permissionCollection;
         this.securityManager = System.getSecurityManager();
+        _codeSource = codeSource;
     }
 
     /**
@@ -125,9 +133,78 @@ public class JasperLoader extends URLClassLoader {
             return clazz;
         }
 
+        String path = name.replace('.', '/') + ".class";
+
+        if (findResource(path) == null) {
+            return _loadFromParent(name, path);
+        }
+
         return findClass(name);
     }
 
+    private Class<?> _loadFromParent(String className, String path) {
+        byte[] classBytes;
+
+        InputStream inputStream = null;
+
+        try {
+
+            if (SecurityUtil.isPackageProtectionEnabled()){
+                inputStream = AccessController.doPrivileged(
+                    (PrivilegedAction<InputStream>)
+                        () -> getParent().getResourceAsStream(path));
+            }
+            else {
+                inputStream = getParent().getResourceAsStream(path);
+            }
+
+            if (inputStream == null) {
+                return null;
+            }
+
+            try (ByteArrayOutputStream byteArrayOutputStream =
+                new ByteArrayOutputStream()) {
+
+                byte buf[] = new byte[1024];
+
+                for (int i = 0; (i = inputStream.read(buf)) != -1; ) {
+                    byteArrayOutputStream.write(buf, 0, i);
+                }
+
+                classBytes = byteArrayOutputStream.toByteArray();
+            }
+        }
+        catch (Exception e) {
+            e.printStackTrace();
+
+            return null;
+        }
+        finally {
+            if (inputStream != null) {
+                try {
+                    inputStream.close();
+                }
+                catch (IOException ioe) {
+                }
+            }
+        }
+
+        Class clazz;
+
+        if (securityManager != null) {
+            ProtectionDomain pd = new ProtectionDomain(
+                _codeSource, permissionCollection);
+
+            clazz = defineClass(
+                className, classBytes, 0, classBytes.length, pd);
+        }
+        else {
+            clazz = defineClass(className, classBytes, 0, classBytes.length);
+        }
+
+        return clazz;
+    }
+
 
     /**
      * Delegate to parent
